text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,558 @@
+| ------------------ | --------------------------- | ------------ | --------------------------------------------------- |
+| `name`             | `string`                    | **Required** | The unqualified name of the union type              |
+| `fqn`              | `string`                    | **Required** | The _jsii_ fully-qualified name of the type         |
+| `kind`             | `@jsii/spec.TypeKind.UNION` | **Required** | The type kind discriminator                         |
+| `assembly`         | `string`                    | **Required** | The name of the assembly containing the union type  |
+| `types`            | `string[]`                  | **Required** | The _jsii_ fully-qualified names of candidate types |
+| `namespace`        | `string`                    |              | The namespace in which the union type is declared   |
+| `docs`             | `@jsii/spec.Docs`           |              | Any documentation attached to the union type        |
+| `locationInModule` | `@jsii/spec.SourceLocation` |              | The source location of the type alias declaration   |
+
+The `jsii` compiler will then be modified in order to locate and process exported _union type_ alias declarations, such
+as:
+
+```ts
+export type UnionType = CandidateA | CandidateB;
+```
+
+In order to ensure ability to generate stable code for _union-like_ types in statically typed languages, the possible
+_candidate types_ of an aliased _type union_ must however be restricted to prevent name collisions:
+
+- all candidates **must** have distinct _unqalified names_:
+  ```ts
+  // Invalid: both types have the unqualified name `Construct`
+  export type InvalidUnion = cdk.Construct | constructs.Construct;
+  ```
+- no candidate type can have a name starting with `ListOf` or `MapOf` followed by an upper-case letter:
+  ```ts
+  // Invalid: ListOfFoo would collide with generated union-like accessors for `Foo[]`
+  export type InvalidUnion = ListOfFoo | Bar;
+  ```
+
+In order to preserve backwards-compatbility, and to retain ability to use this new feature against the current CDK v1
+codebase, the `jsii` compiler will not **require** that _type unions_ are systematically aliased. Similarly,
+`jsii-pacmak` will be modified in such a way that it continues producing the same code whether a _type union_ is aliased
+or not.
+
+Once this done, the `cfn2ts` tool, which is used in the CDK to generate _CloudFormation Resources_ (also known as _L1
+constructs_) will be modified so that it emits aliases for any _type union_ it generates.
+
+Gearing up for the [AWS CDK v2] release, `jsii` will be updated to accept a new option (`--strict=type-unions`) that
+will allow users to opt into **requiring** that any visible _type union_ is aliased. This feature will be gradually
+turned on in the AWS CDK v1 development branch, as all existing _type unions_ are aliased. Once the whole codebase has
```

Comment:
Asides from L1s, the CDK codebase itself almost doesn't have type unions. Can we just get rid of them early? I don't see a huge need for this gradual phase out.

Reply:
Philosophically it does not matter much. I will **need** to gate this with a flag as this is otherwise a breaking change for `jsii` customers!

Whether we fix this flag early or late... doesn't change the flag must be here.